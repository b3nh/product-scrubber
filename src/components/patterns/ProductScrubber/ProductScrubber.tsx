import React, { useEffect, useState } from 'react';

import './product-scrubber.scss';

type dataType = {
  id: string,
  name: string,
  default?: boolean,
  imagePath: string,
}

type ProductScrubberTypes = {
  children?: any,
  onClickItem?: (activeId: string) => void;
  data: Array<dataType>,
  showHoverGrid?: boolean,
  hideName?: boolean,
  hideId?: boolean,
} & typeof defaultProps;

const defaultProps = {
  children: <></>,
  onClickItem: ():any => void (0),
  showHoverGrid: false,
  hideName: false,
  hideId: false,
};

const ProductScrubber = (props: ProductScrubberTypes) => {
  const {
    children,
    onClickItem,
    data,
    showHoverGrid,
    hideName,
    hideId,
  } = props;

  const [loading, setLoading] = useState(true);
  const [activeId, setActiveId] = useState('');
  const [activeName, setActiveName] = useState('');
  const [activeImagePath, setActiveImagePath] = useState('');

  const onHoverItem = (item: dataType) => {
    if (loading) return;
    const { id, name, imagePath } = item;
    setActiveId(id);
    setActiveName(name);
    setActiveImagePath(imagePath);
  };

  const setDefault = () => {
    if (data && data.length) {
      let defaultIndex = data.map((item) => item.default).indexOf(true);
      if (defaultIndex === -1) defaultIndex = 0;
      const { id, name, imagePath } = data[defaultIndex];
      setActiveId(id);
      setActiveName(name);
      setActiveImagePath(imagePath);
    }
  };

  useEffect(() => {
    setDefault();

    const timer = setTimeout(() => {
      setLoading(false);
    }, 500);

    return () => {
      clearTimeout(timer);
    };
  }, []);

  return (
    <div className={`product-scrubber-container ${loading ? 'product-scrubber-container-loading' : 'product-scrubber-container-active'}`}>
      <div className="child-container">
        {children}
      </div>

      {!hideName && <h4>{activeName}</h4>}
      {!hideId && <span>{activeId}</span>}

      {data && data.length > 1
        ? (
          <div className="location">
            {!loading && data.map((item: dataType, i: number) => (
              <div
                key={`item-preview-${i}`}
                className={`item-preview ${item.id === activeId ? 'item-preview-active' : 'item-preview-default'}`}
                style={{ backgroundImage: `url(${item.imagePath})` }}
                onClick={() => onHoverItem(item)}
                onKeyUp={() => onHoverItem(item)}
                aria-hidden
              />
            ))}
          </div>
        ) : null}

      <ul
        onClick={() => onClickItem(activeId)}
        onKeyUp={() => onClickItem(activeId)}
        aria-hidden
      >
        {data && data.length > 1 ? data.map((item: dataType, i: number) => (
          <li
            className={showHoverGrid ? 'show-grid' : 'hide-grid'}
            key={`product-scrubber-image-${i}`}
            onMouseEnter={() => onHoverItem(item)}
            aria-hidden
          />
        )) : null}
      </ul>
      <div
        className={`hover-active-image ${loading ? 'loading' : 'loading-complete'}`}
        style={{ backgroundImage: `url(${activeImagePath})` }}
      />
    </div>
  );
};

ProductScrubber.defaultProps = defaultProps;

export default ProductScrubber;
