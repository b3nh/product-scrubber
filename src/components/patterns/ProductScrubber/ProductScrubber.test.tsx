/* eslint-disable no-undef */

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ProductScrubber from './ProductScrubber';

const TestComponent = (testProps: any) => {
  const { data, onClickItem } = testProps;
  return <ProductScrubber data={data} onClickItem={onClickItem} />;
};

describe('ProductScrubber primary', () => {
  const testProps = {
    data:
      [
        {
          name: 'side',
          default: true,
          imagePath: '../../../stories/assets/side.jpg',
        },
        {
          name: 'front',
          default: false,
          imagePath: '../../../stories/assets/front.jpg',
        },
        {
          name: 'back',
          default: false,
          imagePath: '../../../stories/assets/back.jpg',
        },
        {
          name: 'top',
          default: false,
          imagePath: '../../../stories/assets/top.jpg',
        },
      ],
    onClickItem: jest.fn(),
  };

  // beforeEach(() => {
  //   render(TestComponent(testProps));
  // });

  it('should match snapshot', () => {
    expect(render(TestComponent(testProps))).toMatchSnapshot();
  });
});
