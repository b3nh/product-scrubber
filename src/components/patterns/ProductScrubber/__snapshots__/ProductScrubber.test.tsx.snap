// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductScrubber primary should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="product-scrubber-container product-scrubber-container-loading"
        style="width: 100%; height: 100%;"
      >
        <div
          class="child-container"
        />
        <h4>
          side
        </h4>
        <span />
        <div
          class="location"
        />
        <ul
          aria-hidden="true"
        >
          <li
            aria-hidden="true"
            class="hide-grid"
          />
          <li
            aria-hidden="true"
            class="hide-grid"
          />
          <li
            aria-hidden="true"
            class="hide-grid"
          />
          <li
            aria-hidden="true"
            class="hide-grid"
          />
        </ul>
        <div
          class="hover-active-image loading"
          style="background-image: url(../../../stories/assets/side.jpg);"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="product-scrubber-container product-scrubber-container-loading"
      style="width: 100%; height: 100%;"
    >
      <div
        class="child-container"
      />
      <h4>
        side
      </h4>
      <span />
      <div
        class="location"
      />
      <ul
        aria-hidden="true"
      >
        <li
          aria-hidden="true"
          class="hide-grid"
        />
        <li
          aria-hidden="true"
          class="hide-grid"
        />
        <li
          aria-hidden="true"
          class="hide-grid"
        />
        <li
          aria-hidden="true"
          class="hide-grid"
        />
      </ul>
      <div
        class="hover-active-image loading"
        style="background-image: url(../../../stories/assets/side.jpg);"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
